consensus/hotstuff/verification/staking_verifier.go

```diff
@@ -51,20 +54,24 @@ func (v *StakingVerifier) VerifyVote(signer *flow.Identity, sigData []byte, bloc
 	return nil
 }
 
-// VerifyQC verifies the validity of a single signature on a quorum certificate.
-//
+// VerifyQC checks the cryptographic validity of the QC's `sigData` for the
+// given block. It is the responsibility of the calling code to ensure
+// that all `voters` are authorized, without duplicates. Return values:
+//  - nil if `sigData` is cryptographically valid
+//  - model.ErrInvalidFormat if `sigData` has an incompatible format
+//  - model.ErrInvalidSignature if a signature is invalid
+//  - unexpected errors should be treated as symptoms of bugs or uncovered
+//	  edge cases in the logic (i.e. as fatal)
 // In the single verification case, `sigData` represents a single signature (`crypto.Signature`).
 func (v *StakingVerifier) VerifyQC(signers flow.IdentityList, sigData []byte, block *model.Block) error {
-	// verify the aggregated staking signature
-	msg := MakeVoteMessage(block.View, block.BlockID)
-
-	pks := make([]crypto.PublicKey, 0, len(signers))
-	for _, identity := range signers {
-		pks = append(pks, identity.StakingPubKey)
+	if len(signers) == 0 {
+		return fmt.Errorf("empty list of signers: %w", model.ErrInvalidFormat)
```

Comment:
Whether or not we enter this check is dependent on the global order of all the checks for the QC. Essentially, there are two different checks that both work with the list of `signers`:
1. `signers` should represent a supermajority of consensus participants 
2. the aggregated staking sig should match the aggregated public keys  

This component here only implements check 2. If we check first 1. and then 2., the list of `signers` here should never be empty. However, if we first check 2. and then 1, this component has to handle the case of a malicious QC with empty signer list. 
The order in which we check 1. and 2. is implemented elsewhere. I generally trie as much as possible to reduce pre-conditions on other part of the code base, because it makes the code less brittle:
* adding the check here for `len(signers) == 0` means that the code here works without assuming any order. You can first check 1. and 2. or vice versa. Invalid QCs are detected and rejected. 
* However, without the check for `len(signers) == 0`, this line errors for empty signer list: https://github.com/onflow/flow-go/blob/e61f2934f815e1877cd616965d36753df26b249b/consensus/hotstuff/verification/staking_verifier.go#L74
   * we could require that 1. is checked first, but it would make the code more brittle (as changing the order of checks in some other part of the code would break this code here)
   * alternatively, we could check for an `crypto.invalidInputsError` in line 75: https://github.com/onflow/flow-go/blob/e61f2934f815e1877cd616965d36753df26b249b/consensus/hotstuff/verification/staking_verifier.go#L74-L77 However, then we are unable to distinguish a malformed external input from an internal uncovered edge-case (see [my earlier comment](https://github.com/onflow/flow-go/pull/1830#discussion_r787337554))

Reply:
added the following documentation: https://github.com/onflow/flow-go/blob/44d665742ac284822af7233fc564a766a6837d33/consensus/hotstuff/verification/staking_verifier.go#L75-L89